### DISCUSSION TITLE
Detailed progress and stats tracking over time?

### DISCUSSION THREAD

**Discussion by YannCebron** [didoesdigital/typey-type#85] [Building Typey Type]
(Sorry if this has been discussed elsewhere, couldn't find it)

The detailed statistics and graphs after completing each lesson are a great help to improve, track, and motivate learners.
Unfortunately, there seems to be no "tracking progress over time" - results are _lost_ immediately unless I manually c/p stats into some other location/EXCEL sheet/...

What I'd like to see would be to establish some kind of "account" (could be bound to browser-bound storage even, thus anonymous) that would track ideally _all details_ for _each completed lesson_ and provide me some nice "dashboard" visualizing my learning progress _over time_ - automatically. I understand this is not a small task to do, but it seems kind of "obvious missing next step" to me  - given all the data is already there :) 

Some ideas:
- high-level stats over time
    - for each specific lesson (to see how I improve on specific lesson)
    - overall (to see my WPM go BRRRRR) 
- show progress of words from Gutenberg 10000 word list:
    -  total words shown/typed/memorized
    - number of words with low/high mistype rate
    - top mistype words, maybe even some AI to spot underlying patterns and create automatic custom lesson from it?
    - show words as "bubble diagram" to see most typed words
- visualize performed lessons on a timeline to show what I've been working on and when (with some high-level indicator on progress up/down compared to previous take of this lesson) 
 
I bet there would be dozens more of ideas on what and how to present...

Is this something that is planned or could be considered in the future? Curious what everyone thinks about it. Thanks :)

**Comment by didoesdigital** [comment]
Closest related issue is this one: https://github.com/didoesdigital/typey-type/issues/45

I am suuuuper keen for showing progress over time and would love to make all the viz for it.

The single biggest blocker for me is data storage. Some considerations on that:

- local storage has limited storage space around 5mb and is already used for the list of typed/met words and personal dictionaries so it fills up fast. IndexedDB might be an option to try?
- I am kinda keen to avoid adding a back end or database for account management to simplify security and privacy (and the personal obligation to host and maintain that)
- people sometimes clear their browsing data (history, cookies, and cache) and lose their Typey Type progress in local storage, so I try to make that risk obvious on the progress page and encourage people to make their own copies. This already has friction/low usability for only the typed/met words list so it wouldn't be a great option for adding more stats data to manage
- I want people to own/have access to all their data

I would guestimate something like storing speed and accuracy per completed lesson for an extremely diligent student would only stack up to 2MB over a year, but more detailed stats on mistyped words or progress of typed/met words over time could be much more so we'd want to be strategic about what to record.

**Comment by YannCebron** [comment]
Thanks, that sounds promising. Agree on all points about data storage vs. dedicated account/login. I'd expect most people have  only one computer on which they use/practice Steno (special keyboard!) anyway; and the data export could probably be expanded to dump/import all these stats as you said.
